pkg/apis/v1/nodepool.go

```diff
@@ -0,0 +1,313 @@
+// launch nodes in response to pods that are unschedulable. A single nodepool
+// is capable of managing a diverse set of nodes. Node properties are determined
+// from a combination of nodepool and pod scheduling constraints.
+type NodePoolSpec struct {
+	// Template contains the template of possibilities for the provisioning logic to launch a NodeClaim with.
+	// NodeClaims launched from this NodePool will often be further constrained than the template specifies.
+	// +required
+	Template NodeClaimTemplate `json:"template"`
+	// Disruption contains the parameters that relate to Karpenter's disruption logic
+	// +kubebuilder:default={"consolidationPolicy": "WhenUnderutilized", "expireAfter": "720h"}
+	// +kubebuilder:validation:XValidation:message="consolidateAfter cannot be combined with consolidationPolicy=WhenUnderutilized",rule="has(self.consolidateAfter) ? self.consolidationPolicy != 'WhenUnderutilized' || self.consolidateAfter == 'Never' : true"
+	// +kubebuilder:validation:XValidation:message="consolidateAfter must be specified with consolidationPolicy=WhenEmpty",rule="self.consolidationPolicy == 'WhenEmpty' ? has(self.consolidateAfter) : true"
+	// +optional
+	Disruption Disruption `json:"disruption"`
+	// Limits define a set of bounds for provisioning capacity.
+	// +optional
+	Limits Limits `json:"limits,omitempty"`
+	// Weight is the priority given to the nodepool during scheduling. A higher
+	// numerical weight indicates that this nodepool will be ordered
+	// ahead of other nodepools with lower weights. A nodepool with no weight
+	// will be treated as if it is a nodepool with a weight of 0.
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=100
+	// +optional
+	Weight *int32 `json:"weight,omitempty"`
+}
+
+type Disruption struct {
+	// ConsolidateAfter is the duration the controller will wait
+	// before attempting to terminate nodes that are underutilized.
+	// Refer to ConsolidationPolicy for how underutilization is considered.
+	// +kubebuilder:validation:Pattern=`^(([0-9]+(s|m|h))+)|(Never)$`
+	// +kubebuilder:validation:Type="string"
+	// +kubebuilder:validation:Schemaless
+	// +optional
+	ConsolidateAfter *NillableDuration `json:"consolidateAfter,omitempty"`
+	// ConsolidationPolicy describes which nodes Karpenter can disrupt through its consolidation
+	// algorithm. This policy defaults to "WhenUnderutilized" if not specified
+	// +kubebuilder:default:="WhenUnderutilized"
+	// +kubebuilder:validation:Enum:={WhenEmpty,WhenUnderutilized}
+	// +optional
+	ConsolidationPolicy ConsolidationPolicy `json:"consolidationPolicy,omitempty"`
+	// ExpireAfter is the duration the controller will wait
+	// before terminating a node, measured from when the node is created. This
+	// is useful to implement features like eventually consistent node upgrade,
+	// memory leak protection, and disruption testing.
+	// +kubebuilder:default:="720h"
+	// +kubebuilder:validation:Pattern=`^(([0-9]+(s|m|h))+)|(Never)$`
```

Comment:
Can we do a pass to see if we are on a new-enough version now that we can rely more heavily on CEL validation if it's appropriate